File: Library/KMP.py
def failure_function(pattern):
    m=len(pattern)
    failure_array=[None]*m
    failure_array[0]=0
    k=0
    for q in range(1,m):
        while k>0 and pattern[k]!=pattern[q]:
            k=failure_array[k-1]
        if pattern[k]==pattern[q]:
            k=k+1
        failure_array[q]=k
    return failure_array

def KMP(text,pattern):
    F=failure_function(pattern)
    len_text=len(text)
    len_pattern=len(pattern)
    q=0
    for i in range(len(text)):
        while q>0 and pattern[q]!=text[i]:
            q=F[q]
        if text[i]==pattern[q]:
            q=q+1

        if q==len_pattern:
            print("match at :",i-(len_pattern-1)," index")
            q=F[q-1]

t=input("t")
p=input("p")
KMP(t,p)
        
        

File: Library/KMP.cpp
#include <iostream> 
#include <string>

using namespace std;

void failureFunction (string pattern, int *charShift) 
{
	charShift[0] = 0;
	int k = 0;
	for (int i=1; i<pattern.length(); i++) {
		while (k>0 && pattern[k] != pattern[i]) {
			k = charShift[k-1];
		}
		if (pattern[k] == pattern[i]) {
			k++;
		}
		charShift[i] = k;
	}
}

void KMPSearch (string text, string pattern) {
	int patternLen = pattern.length();
	int textLen = text.length();
	int charShift [patternLen];
	failureFunction (pattern, charShift);
	int q=0;
	for(int i=0;i<text.length();i++){
		while (q>0 && pattern[q]!=text[i]){
			q=charShift[q];
		}
		if (text[i]==pattern[q]){
			q++;
		}
		if (q==patternLen){
			cout<<"Match found at : "<<(i-(patternLen-2))<<" Index";
			q=charShift[q-1];
			cout<<endl;
		}
	}

}

int main (void) {
	string text, pattern;
	getline (cin, text);
	getline (cin, pattern);
	KMPSearch (text, pattern);
	return 500;
}